File: disciplinas/estrutura-de-dados-I/implementacoes/listas-encadeadas/python/lista_encadeada.py
# Disciplina: Estrutura de Dados I
# Assunto: Listas Encadeadas
# Descricao: Implementacao de lista encadeada simples em Python com operacoes basicas

class ListaEncadeada:
    class No:
        def __init__(self, valor):
            self.valor = valor
            self.proximo = None

    def __init__(self):
        self.inicio = None
        self.tamanho = 0

    def inserir_posicao(self, valor, posicao):
        if posicao < 0 or posicao > self.tamanho:
            raise IndexError("Posicao invalida")

        novo_no = self.No(valor)

        if posicao == 0:
            novo_no.proximo = self.inicio
            self.inicio = novo_no
            self.tamanho += 1
            return

        anterior = self.no_na_posicao(posicao - 1)
        novo_no.proximo = anterior.proximo
        anterior.proximo = novo_no
        self.tamanho += 1

    def remover_posicao(self, posicao):
        if posicao < 0 or posicao >= self.tamanho:
            raise IndexError("Posicao invalida")

        if posicao == 0:
            valor_removido = self.inicio.valor
            self.inicio = self.inicio.proximo
            self.tamanho -= 1
            return valor_removido

        anterior = self.no_na_posicao(posicao - 1)
        removido = anterior.proximo
        anterior.proximo = removido.proximo
        self.tamanho -= 1
        return removido.valor

    def buscar_valor(self, valor):
        atual = self.inicio
        posicao = 0

        while atual is not None:
            if atual.valor == valor:
                return posicao
            atual = atual.proximo
            posicao += 1

        return -1

    def no_na_posicao(self, posicao):
        atual = self.inicio
        for _ in range(posicao):
            atual = atual.proximo
        return atual

    def __str__(self):
        valores = []
        atual = self.inicio

        while atual is not None:
            valores.append(str(atual.valor))
            atual = atual.proximo

        return "[" + " -> ".join(valores) + "]"

    def liberar(self):
        self.inicio = None
        self.tamanho = 0

    def get_tamanho(self):
        return self.tamanho

    @staticmethod
    def main():
        lista = ListaEncadeada()

        try:
            lista.inserir_posicao(10, 0)
            lista.inserir_posicao(20, 1)
            lista.inserir_posicao(40, 2)
            print("Lista inicial:", lista)

            lista.inserir_posicao(30, 2)
            print("Apos inserir 30 na posicao 2:", lista)

            removido = lista.remover_posicao(1)
            print("Valor removido da posicao 1:", removido)
            print("Lista apos remocao:", lista)

            posicao = lista.buscar_valor(30)
            print("Valor 30 encontrado na posicao:", posicao)

        except IndexError as e:
            print("Erro:", e)

if __name__ == "__main__":
    ListaEncadeada.main()

File: disciplinas/estrutura-de-dados-I/implementacoes/listas-encadeadas/java/ListaEncadeada.java
// Disciplina: Estrutura de Dados I
// Assunto: Listas Encadeadas
// Descricao: Implementacao de lista encadeada simples em Java com operacoes basicas

public class ListaEncadeada {
    private No inicio;
    private int tamanho;

    public ListaEncadeada() {
        this.inicio = null;
        this.tamanho = 0;
    }

    private static class No {
        int valor;
        No proximo;

        No(int valor) {
            this.valor = valor;
            this.proximo = null;
        }
    }

    public void inserirPosicao(int valor, int posicao) throws IndexOutOfBoundsException {
        if (posicao < 0 || posicao > tamanho) {
            throw new IndexOutOfBoundsException("Posicao invalida");
        }

        No novoNo = new No(valor);

        if (posicao == 0) {
            novoNo.proximo = inicio;
            inicio = novoNo;
            tamanho++;
            return;
        }

        No anterior = noNaPosicao(posicao - 1);
        novoNo.proximo = anterior.proximo;
        anterior.proximo = novoNo;
        tamanho++;
    }

    public int removerPosicao(int posicao) throws IndexOutOfBoundsException {
        if (posicao < 0 || posicao >= tamanho) {
            throw new IndexOutOfBoundsException("Posicao invalida");
        }

        if (posicao == 0) {
            int valorRemovido = inicio.valor;
            inicio = inicio.proximo;
            tamanho--;
            return valorRemovido;
        }

        No anterior = noNaPosicao(posicao - 1);
        No removido = anterior.proximo;
        anterior.proximo = removido.proximo;
        tamanho--;
        return removido.valor;
    }

    public int buscarValor(int valor) {
        No atual = inicio;
        int posicao = 0;

        while (atual != null) {
            if (atual.valor == valor) {
                return posicao;
            }
            atual = atual.proximo;
            posicao++;
        }

        return -1;
    }

    private No noNaPosicao(int posicao) {
        No atual = inicio;
        for (int i = 0; i < posicao; i++) {
            atual = atual.proximo;
        }
        return atual;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");

        No atual = inicio;
        while (atual != null) {
            sb.append(atual.valor);
            if (atual.proximo != null) {
                sb.append(" -> ");
            }
            atual = atual.proximo;
        }

        sb.append("]");
        return sb.toString();
    }

    public void liberar() {
        inicio = null;
        tamanho = 0;
    }

    public int getTamanho() {
        return tamanho;
    }

    public static void main(String[] args) {
        ListaEncadeada lista = new ListaEncadeada();

        try {
            lista.inserirPosicao(10, 0);
            lista.inserirPosicao(20, 1);
            lista.inserirPosicao(40, 2);
            System.out.println("Lista inicial: " + lista);

            lista.inserirPosicao(30, 2);
            System.out.println("Apos inserir 30 na posicao 2: " + lista);

            int removido = lista.removerPosicao(1);
            System.out.println("Valor removido da posicao 1: " + removido);
            System.out.println("Lista apos remocao: " + lista);

            int posicao = lista.buscarValor(30);
            System.out.println("Valor 30 encontrado na posicao: " + posicao);

        } catch (IndexOutOfBoundsException e) {
            System.out.println("Erro: " + e.getMessage());
        }
    }
}